lib/src/main/java/com/knuddels/jtokkit/TokenEncoder.java

```diff
@@ -101,13 +103,32 @@ public int addTokensAndGetCount(int maxTokenCount, boolean keepEncodings, byte[]
             return 1;
         } else {
             int length = match.length();
-            return addTokensAndGetCount(maxTokenCount, keepEncodings, out, ranks, match, length);
+            if (length < VERY_LARGE_TOKENIZER_BYTE_THRESHOLD) {
+                return addTokensAndGetCountSmall(maxTokenCount, keepEncodings, out, ranks, match, length);
+            } else {
+                return addTokensAndGetCountLarge(this, maxTokenCount, keepEncodings, out, match, length);
+            }
         }
     }
 
-    private int addTokensAndGetCount(int maxTokenCount, boolean keepEncodings, List<Integer> out, List<Integer> ranks, ImmutableByteArray match, int length) {
-        int validRanks = initRanks(match, length, ranks);
-        int tokenCount = mergeBytesAndGetTokenCount(match, length, ranks, validRanks);
+    private int addTokensAndGetCountSmall(int maxTokenCount, boolean keepEncodings, List<Integer> out, List<Integer> ranks, ImmutableByteArray match, int length) {
+        assert length > 1 : "Already filtered out";
+        ranks.clear();
+
+        int validRanks = 0;
+        int minRankIndex = -1;
```

Comment:
since we're already iterating to calculate the initial ranks, let's get the first minimum and  validRanks as well (avoiding yet another iteration)